### DISCUSSION TITLE
why use assign and timing of invoke

### DISCUSSION THREAD

**Discussion by shmdhussain** [statelyai/xstate#1850] [Q&A]
Hi David,

This week I watched the xstate course on frontendMasters.

I have few doubts.

--first one is:

Why we need to use assign when we can directly do the same thing in action function.

I heard you are explaining this , but I am unable to understand the mutable and immutable context. could you point me to some pointers or explain me

------------------------------------------
--Second One is: 

When will the invoke triggered, after entering into the state?

------------------------------------------

thanks in advance,

**Comment by Andarist** [comment]
> Why we need to use assign when we can directly do the same thing in action function.

You can't. `assign` is used to manipulate machine's context value which is treated as immutable for a variety of reasons. While you probably can get away with:
```
function someAction(ctx) {
  ctx.foo.bar = 'newValue'
}
```
you won't be able to manipulate the context correctly for any top-level key, so this won't work as expected:
```js
function someAction(ctx) {
  ctx.foo = 'newValue'
}
```

Either way - we do not recommend direct mutations to manipulate the context value because it makes certain things impossible (like going back in history in the visualizer etc).

> When will the invoke triggered, after entering into the state?

Yes, always after entering the state.
